File: problem_solving_with_DSA/stack/C/postFixEval.c
#include "stack.h"

/**
 * postFixEval - Evaluate postFix expression
 * 
 * @str - Postfix
 * Return - the result of postfix expression 
 */
int postFixEval(char *str)
{
    int i;
    
    for (i = 0; str[i]; i++)
    {
        if (str[i] >= '0' && str[i] <= '9')
            push(str[i]-'0'); // convert to integer
        else
        {
            int x = pop();
            int y = pop();
            
            switch (str[i])
            {
            case '+':
                push(y+x);
                break;
            case '-':
                push(y-x);
                break;
            case '*':
                push(y*x);
                break;
            case '/':
                push(y/x);
                break;
            }
        }
    }
    return pop();
}


int main(void)
{
    char *postFix = "82+";
    // postFix = "825+*5/";
    postFix = "525+*5/";    
    int res = postFixEval(postFix);
    
    printf("%d\n", res);

    return (0);
}

File: problem_solving_with_DSA/stack/Python/postFixEval.py
#!/usr/bin/env python3
"""Evaluate postfix expression"""

def postFixEval(str: str) -> int:
    """Evaluate postfix expression
    
    Args:
        str (str): postfix expression
        
    Return:
        (int): Result of the expression
    """
    stack = []
    for ch in str:
        if (ch >= '0' and ch <= '9'):
            stack.append(int(ch))
        else:
            x = stack.pop()
            y = stack.pop()
            
            match ch:
                case '+':
                    stack.append(y+x)
                    break
                case '-':
                    stack.append(y-x)
                    break
                case '*':
                    stack.append(y*x)
                    break
                case '/':
                    stack.append(y/x)
                    break
    return stack.pop()
    

if __name__ == '__main__':
    postfix = "82+"
    
    res = postFixEval(postfix)
    print(res)